### ISSUE TITLE
CKEditor 5 dialogs from within a Backdrop (jquery.ui) dialog not working

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#6351]
## Description of the bug

It's a tricky problem.

When editing block content via layout form, which opens the editor in a dialog, the dialog traps the focus, so opening a CKEditor modal fails

## Steps To Reproduce

1. Create a custom block in a layout
2. Insert an image (the Backdrop image dialog *is not the culprit here*)
3. Try to set the alternative text for the image

## Actual behavior

For a _moment_ the alt text dialog seems to open, but closes immediately after that.
It's not possible to input any alternative text for **inline images in blocks** with this context toolbar item.

![editor-dialog-detail](https://github.com/backdrop/backdrop-issues/assets/842066/b9426985-7381-4289-bb9c-75c3b9db9b9c)

## Expected behavior

Dialogs should work.

## Additional information

Seems to be some conflict between jquery.ui dialog (Backdrop), which tries to trap the focus it seems, and the CKEditor 5 dialog, which doesn't get the focus, so form elements like text input don't work.

It might be possible to catch that aggressive `focusin` handling by preventing propagation under some circumstances. Or override _allowInteraction in the jquery widget. Or we could drop that form element, as alt text can still get set via image dialog.

**Comment by indigoxela** [comment]
The Drupal v4 module struggles with the same issue, but their approach doesn't fix our problem with balloon toolbars. But with some adaptions it seems to work.

How to test:

- Make sure, CKE5 is enabled for the format
- Go to any layout form, like /admin/structure/layouts/manage/default
- Add a custom block
- Add an inline image in that block
- Try to set the alt text via balloon toolbar (the iconbar above the image)

That should work now, and the console error re `Uncaught CKEditorError: contextualballoon-remove-view-not-exist` should be gone.

**Comment by quicksketch** [comment]
Thanks @indigoxela for investigating this. I went down a Google search rabbit hole and ran across https://bugs.jqueryui.com/ticket/9087/, which seems to describe the situation we're facing pretty clearly. It also recommends doing basically what you've suggested as the "official" solution to this problem.

In that jQuery UI issue, it's interesting that they describe the problem affecting CKEditor 4. It's curious that we didn't see this problem before, or did we just not have a report on it?

A couple suggestions on the implementation:

- Could we move this into `modules/ckeditor/js/ckeditor.js`, since we're going to need this work-around pretty much 100% of the time. Or will dependency loading be problematic?
- Include a link to https://bugs.jqueryui.com/ticket/9087/ and/or https://api.jqueryui.com/dialog/#method-_allowInteraction in the inline documentation.

**Comment by herbdool** [comment]
I'm setting the milestone to 1.27.0 so it can be fixed along with introducing cke5.

**Comment by indigoxela** [comment]
> It's curious that we didn't see this problem before, or did we just not have a report on it?

The problem exists, also with v4, we just didn't get any report on this, yet, I guess. BTW, other features in v4 don't work in dialogs as well, like maximize.

There's a contrib module hit by this bug: video_filter. It provides a CKE4 dialog, which is unusable in layout forms (on block edit) because of the lack of focus.

> Could we move this into modules/ckeditor5/js/ckeditor5.js

I tried that, but unfortunately it didn't work ("i is not a constructor")...

> we're going to need this work-around pretty much 100% of the time

Only for dialogs. The "normal" edit experience in node forms and such isn't affected. Only when the editor gets loaded inside a jquery.ui dialog.
I'm not sure, if the current approach is ideal - it's maybe not generic enough. But I wouldn't know, how to make this more generic, so the jquery.ui core library could provide something OOTB.

**Comment by indigoxela** [comment]
Updated the PR, added the jquery.ui issue link, rebased.
After checking a CKE5 demo with a dialog, I verified, that everything, also the dialog itself gets the "ck" class, so I simplified the check.

Any hints for further improvement are welcome.

**Comment by herbdool** [comment]
RTBC. Code looks good and tested by adding alt text to an image inserting into a custom block in a layout. (whew)

**Comment by quicksketch** [comment]
Tested and confirmed it fixes balloon toolbar use in CKEditor 5. I added some additional comments to the PR now that I have a better understanding of what the problem is.

We probably don't see a lot of issues with CKEditor 4 because this issue only comes up when using CKEditor 4 modals (for example when editing a table's properties), rather than Backdrop-provided modals (such as editing a link or inserting an image). CKEditor 5 has a unique problem in that the "balloon" toolbar is in the root of the DOM. Since that is outside the modal dialog DIV, jQuery UI blocks interaction with it. The patch basically says, "even if this element is outside the modal, allow interaction with it".

Anyway, RTBC +2. Thanks @indigoxela and @herbdool!
